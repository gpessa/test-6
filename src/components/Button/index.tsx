import PropTypes, { InferProps } from "prop-types";
import React from "react";

import styles from "./styles.module.scss";

const Button: React.FC<Props> = ({
  className,
  children,
  variant,
  type,
  ...props
}) => (
  <button
    type={type as React.ButtonHTMLAttributes<HTMLButtonElement>["type"]}
    className={`${styles.button} ${className} ${variant &&
      styles[`button--${variant}`]}`}
    {...props}
  >
    {children}
  </button>
);

const propTypes = {
  type: PropTypes.oneOf(["button", "reset", "submit", null]),
  variant: PropTypes.oneOf(["primary", "secondary", "link", null])
};

const defaultProps = {
  type: "button",
  variant: "primary"
};

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;

type Props = InferProps<typeof propTypes> & React.HTMLProps<HTMLButtonElement>;

export default Button;
