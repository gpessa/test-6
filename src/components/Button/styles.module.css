:local(.element) {
  display: inline-block;
  font-weight: 400;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  user-select: none;
  border: 1px solid transparent;
  padding: 0.375rem 0.75rem;
  font-size: 1rem;
  line-height: 1.5;
  text-transform: uppercase;
  letter-spacing: 2px;
}

:local(.primary) {
  background-color: var(--primary);
  color: var(--white);
}

:local(.secondary) {
  background-color: var(--secondary);
  color: var(--white);
}
