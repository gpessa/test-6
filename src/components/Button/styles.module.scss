@import "../../theme/variables.scss";

:local(.button) {
  display: inline-block;
  font-weight: 400;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  user-select: none;
  border: 1px solid transparent;
  padding: 0.375rem 0.75rem;
  font-size: 1rem;
  line-height: 1.5;
  text-transform: uppercase;
  letter-spacing: 2px;
  text-decoration: none;

  &:disabled {
    cursor: not-allowed;
  }

  &:focus {
    box-shadow: 0 0 8px $secondary             !important;        
  }
}

:local(.button--primary) {
  background-color: $primary;
  color: $white;

  &:disabled {
    background-color: lighten($primary, 10%);
  }

}

:local(.button--secondary) {
  background-color: $secondary;
  color: $white;

  &:disabled {
    background-color: lighten($secondary, 3%);
  }
}
