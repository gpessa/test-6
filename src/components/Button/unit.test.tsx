import React from "react";
import Button from "./";
import { mount, shallow } from "enzyme";
import renderer from "react-test-renderer";

describe("<Button>", () => {
  it("Render correctly Button component", () => {
    const element = renderer.create(<Button>Sumbit</Button>).toJSON();
    expect(element).toMatchSnapshot();
  });

  it("Should have type=button by default", () => {
    const type = mount(<Button>Sumbit</Button>)
      .find("button")
      .getDOMNode()
      .getAttribute("type");

    expect(type).toBe("button");
  });

  it("Should show the type if passed one", () => {
    const type = mount(<Button type="submit">Sumbit</Button>)
      .find("button")
      .getDOMNode()
      .getAttribute("type");

    expect(type).toBe("submit");
  });

  it("Should call onClick callback", done => {
    mount(<Button onClick={() => done()}>Sumbit</Button>).simulate("click");
  });

  it("Should apply variant class", () => {
    const wrapper = shallow(<Button variant="secondary">Sumbit</Button>);
    expect(wrapper.hasClass("button--secondary")).toEqual(true);
  });

  it('Should default to variant="primary"', () => {
    const wrapper = shallow(<Button>Sumbit</Button>);
    expect(wrapper.hasClass("button--primary")).toEqual(true);
  });
});
