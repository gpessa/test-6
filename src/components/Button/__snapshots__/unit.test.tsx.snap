// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> Render correctly Button component 1`] = `
<button
  className="button undefined button--primary"
  type="button"
>
  Title
</button>
`;
